scripts/zkai_attention_kv_d16_quantized_softmax_receipt_gate.py

```
@@ -0,0 +1,586 @@
+    per_step_denominators: list[int] = []
+    grouped = rows_by_step(rows)
+    if sorted(grouped) != list(range(len(outputs))):
+        raise QuantizedSoftmaxReceiptGateError("step grouping drift")
+
+    for step_index in sorted(grouped):
+        step_rows = grouped[step_index]
+        if not step_rows:
+            raise QuantizedSoftmaxReceiptGateError("empty step rows")
+        selected_scores = {row["selected_score"] for row in step_rows}
+        if len(selected_scores) != 1:
+            raise QuantizedSoftmaxReceiptGateError("per-step selected score drift")
```

Ruff (0.15.12)

[warning] Avoid specifying long messages outside the exception class

(TRY003)